### PR TITLE
feat: broker client systemcheck endpoint

### PR DESCRIPTION
**Pull request by adrukh**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/broker/blob/master/.github/CONTRIBUTING.md) rules

#### What does this PR do?

Adds a `/systemcheck` endpoint to the broker client, which pokes a configured validation URL, asserting network reachability and some credentials correctness of the broker client configuration. Main purpose is to assist in troubleshooting of connectivity / application issues.

Environment variables configuring the systemcheck:
`BROKER_CLIENT_VALIDATION_URL` - the URL to test
`BROKER_CLIENT_VALIDATION_METHOD` - the HTTP method, default `GET`
`BROKER_CLIENT_VALIDATION_TIMEOUT_MS` - the HTTP request timeout, default 5000ms
`BROKER_CLIENT_VALIDATION_AUTHORIZATION_HEADER` - the `Authorization` header value to be used on the request, default none
`BROKER_CLIENT_VALIDATION_BASIC_AUTH` - the basic auth creds to be base64-encoded and used for the `Authorization` header value, default none

### CODE REVIEW

File: lib/client/index.js [hunk 1]
```diff
@@ -36,7 +39,34 @@ module.exports = ({ port = null, config = {}, filters = {} }) => {
...
+    }
+
+    // add the internal validation URL to the response data and check it
+    data.brokerClientValidationUrl = logger.sanitise(config.brokerClientValidationUrl);
+    request(config.brokerClientValidationUrl, { timeout: 5000 }, (error, response) => {
+      if (error) {
+        data.brokerClientValidationUrlSuccess = false;
+        data.error = error.message;
+        data.ok = false;
+        return res.status(500).json(data);
```
Comment (joshje) [inline lib/client/index.js]: Do we want to know about `isConnOpen` here? I don't see it being added to `data`, so this potentially important info will not be in the response.

File: lib/client/index.js [hunk 2]
```diff
@@ -36,7 +39,34 @@ module.exports = ({ port = null, config = {}, filters = {} }) => {
...
+      }
+
+      const goodStatusCode = ((response && response.statusCode) < 400);
+      if (!goodStatusCode) {
+        data.brokerClientValidationUrlSuccess = false;
+        data.brokerClientValidationUrlStatusCode = response && response.statusCode;
+        data.error = 'Internal validation URL status code is not 2xx or 3xx';
+        data.ok = false;
+        return res.status(500).json(data);
```
Comment (joshje) [inline lib/client/index.js]: Same as above regarding `isConnOpen` being sent back

File: lib/log.js [hunk 3]
```diff
@@ -35,6 +35,14 @@ function sanitise(raw) {
     raw = sanitiseConfigVariable(raw, 'GITLAB_TOKEN');
   }
 
+  if (config.JIRA_USERNAME) {
```
Comment (joshje) [inline lib/log.js]: Oooh, I didn't realise we needed to add these to sanitise here!

How about looking for `Object.keys(config)` containing `USERNAME` or `PASSWORD` to catch them automatically?

File: lib/client/index.js (hunk 1 above)
Comment (adrukh) [reply lib/client/index.js]: It appears on the `websocketConnectionOpen` key once `data` is declared.

The code is slightly convoluted as I wanted to stick to the callback-way of calling `request` like in the rest of the codebase. Open to suggestions on how to improve, what do you think would be easier to grok?

File: lib/log.js (hunk 3 above)
Comment (adrukh) [reply lib/log.js]: Some creds include `TOKEN` and not `USERNAME`. Agree this is too explicit, but didn't want to focus on this part just yet :)

File: README.md [hunk 4]
```diff
@@ -249,7 +249,13 @@ Note that `BROKER_CLIENT_URL` now has the HTTPS scheme.
 
 The broker exposes an endpoint at `/healthcheck`, which can be used to monitor the health of the running application. This endpoint returns `200 OK` status code when the application is healthy, and will return a JSON object containing `ok: true`.
 
-To change the location of this endpoint, you can specify an alternative path via an environment variable:
+In the case of the broker client, this endpoint reports on two more checks:
+* Check on the status of the broker websocket connection. If the connection is not open, the healthcheck fails.
+* [Optional] Make a request to a specified validation URL (`BROKER_CLIENT_VALIDATION_URL`, `BROKER_CLIENT_VALIDATION_AUTHORIZATION_HEADER` and `BROKER_CLIENT_VALIDATION_METHOD`environment variables). If the request fails (error or 4xx / 5xx status code), the healthcheck fails.
+
+In case of failure, the healtcheck endpoint returns a `500` status code, and the response body elaborates on the failure reasons.
```
Comment (adrukh) [inline README.md]: typo

File: client-templates/jira/.env.sample [hunk 5]
```diff
@@ -7,13 +7,18 @@ JIRA_USERNAME=
 # your personal password to your Jira Server account
 JIRA_PASSWORD=
 
+# your Jira Server hostname, i.e. jira.yourdomain.com
+JIRA_HOSTNAME=
```
Comment (adrukh) [inline client-templates/jira/.env.sample]: breaking change for current jira broker users!

File: dockerfiles/bitbucket-server/Dockerfile [hunk 6]
```diff
@@ -59,10 +59,13 @@ ENV ACCEPT accept.json
...
 ENV BROKER_HEALTHCHECK_PATH /healthcheck
 
+# Bitbucket server validation url, checked by broker client healthcheck endpoint
+ENV BROKER_CLIENT_VALIDATION_URL https://$BITBUCKET_USERNAME:$BITBUCKET_PASSWORD@$BITBUCKET/rest/api/1.0/users
+
 
 EXPOSE $PORT
 
-HEALTHCHECK --interval=10s --timeout=1s \
+HEALTHCHECK --interval=60s --timeout=10s \
```
Comment (adrukh) [inline dockerfiles/bitbucket-server/Dockerfile]: Remove docker healthchecks, users should rely on `/healthcheck` instead

File: lib/client/index.js [hunk 7]
```diff
@@ -36,7 +39,47 @@ module.exports = ({ port = null, config = {}, filters = {} }) => {
...
+    if (config.brokerClientValidationAuthorizationHeader) {
+      validationRequestHeaders.authorization = config.brokerClientValidationAuthorizationHeader;
+    }
+
+    // make the internal validation request
+    request({
+      url: config.brokerClientValidationUrl,
+      headers: validationRequestHeaders,
+      method: config.brokerClientValidationMethod || 'GET',
+      timeout: 5000,
```
Comment (adrukh) [inline lib/client/index.js]: make a config value

File: lib/client/index.js [hunk 8]
```diff
@@ -36,7 +39,47 @@ module.exports = ({ port = null, config = {}, filters = {} }) => {
...
+    data.brokerClientValidationUrl = logger.sanitise(config.brokerClientValidationUrl);
+    const validationRequestHeaders = {
+      'user-agent': 'Snyk Broker client ' + version,
+    };
+    if (config.brokerClientValidationAuthorizationHeader) {
+      validationRequestHeaders.authorization = config.brokerClientValidationAuthorizationHeader;
+    }
+
+    // make the internal validation request
+    request({
```
Comment (adrukh) [inline lib/client/index.js]: Make sure `request` follows redirects by default

File: lib/client/index.js [hunk 9]
```diff
@@ -36,7 +39,47 @@ module.exports = ({ port = null, config = {}, filters = {} }) => {
...
+      timeout: 5000,
+    }, (error, response) => {
+      if (error) {
+        data.brokerClientValidationUrlSuccess = false;
+        data.error = error.message;
+        data.ok = false;
+        return res.status(500).json(data);
+      }
+
+      const goodStatusCode = ((response && response.statusCode) < 400);
```
Comment (adrukh) [inline lib/client/index.js]: `/^2/.test(statusCode)`?

File: lib/client/index.js (hunk 1 above)
Comment (joshje) [reply lib/client/index.js]: Once I uncollapsed the diff on GH I spotted the declaration! I think the confusing part is actually that `status` refers specifically to the result of `isConnOpen`.

What do you think about calling it `isConnOpenStatus`, or query `isConnOpen` boolean at the point we want to determine the status code?

File: client-templates/jira/.env.sample (hunk 5 above)
Comment (joshje) [reply client-templates/jira/.env.sample]: This won't work in cases where the username has an `@` in it (e.g. an email address). No amount of encoding will help either. This needs to become similar to how we handle the `auth` object as defined in the rules:

```
https://github.com/snyk/broker/blob/master/client-templates/jira/accept.json.sample#L22
```

This is so we can use request's auth handler:
https://github.com/snyk/broker/blob/c0418a3b388e33aa3f2458d3275ae05d1633ce4b/lib/filters/index.js#L137

File: README.md [hunk 10]
```diff
@@ -249,7 +249,13 @@ Note that `BROKER_CLIENT_URL` now has the HTTPS scheme.
```
Comment (joshje) [inline README.md]: If we end up going with `JIRA_HOSTNAME` then make sure to update line 190 above

File: lib/client/index.js (hunk 1 above)
Comment (adrukh) [reply lib/client/index.js]: Naming is hard! `isX` is a predicate, so `isConnOpenStatus` doesn't sound proper to me. Maybe `isConnStatusOpen`?

The split flow here is to accommodate the optionality of the validation url, and use `request` in a callback fashion if it is present. Sort of older-school JS... Didn't want to add promise-based `request` lib or similar, or introduce plain `Promise`s just for this flow.

File: client-templates/jira/.env.sample (hunk 5 above)
Comment (adrukh) [reply client-templates/jira/.env.sample]: WOW

I didn't think of this at all!

I did however test this for real with our own jira creds, which included an email address as the username - and it worked! So I'd like to keep it this way, or maybe defer jira validation logic to a later time.

File: README.md (hunk 10 above)
Comment (adrukh) [reply README.md]: Totally!

**Comment by adrukh** [comment]
I'm going to split the validation into its own endpoint, not to mix healthcheck logic (broker client is responsible for the websocket connection) and systemcheck-like logic (incorrect creds will prevent the broker client from working properly, but it is not the broker client's "fault").

**Comment by adrukh** [comment]
@joshje this is now updated with proper handling of basic auth, thanks for bringing this up!

Please review again. 🙏

**Pull request merged**